### PR TITLE
[esp_http_client] Include port in Host field. (IDFGH-1373)

### PR DESCRIPTION
**Pull request by viettrungluu**
Also fix some memory leaks in some failure cases.

https://github.com/espressif/esp-idf/issues/3628

### CODE REVIEW

File: components/esp_http_client/esp_http_client.c [hunk 1]
```diff
@@ -11,8 +11,11 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+//
+// Modifications copyright 2019 Tricot Inc.
```
Comment (viettrungluu) [inline components/esp_http_client/esp_http_client.c]: Typically additional notice is required by the Apache license, though the license grant in the contributor agreement is broader. (That said, it's a license grant, not a copyright assignment.)

File: components/esp_http_client/esp_http_client.c [hunk 2]
```diff
@@ -660,31 +677,8 @@ esp_err_t esp_http_client_set_url(esp_http_client_handle_t client, const char *u
         ESP_LOGE(TAG, "Error parse url %s", url);
         return ESP_ERR_INVALID_ARG;
     }
-    if (client->connection_info.host) {
-        old_host = strdup(client->connection_info.host);
```
Comment (viettrungluu) [inline components/esp_http_client/esp_http_client.c]: This didn't check the failure case.

File: components/esp_http_client/esp_http_client.c [hunk 3]
```diff
@@ -660,31 +677,8 @@ esp_err_t esp_http_client_set_url(esp_http_client_handle_t client, const char *u
...
-    }
     old_port = client->connection_info.port;
 
-    if (purl.field_data[UF_HOST].len) {
-        http_utils_assign_string(&client->connection_info.host, url + purl.field_data[UF_HOST].off, purl.field_data[UF_HOST].len);
-        HTTP_MEM_CHECK(TAG, client->connection_info.host, return ESP_ERR_NO_MEM);
```
Comment (viettrungluu) [inline components/esp_http_client/esp_http_client.c]: old_host was leaked here.

File: components/esp_http_client/esp_http_client.c [hunk 4]
```diff
@@ -700,37 +694,57 @@ esp_err_t esp_http_client_set_url(esp_http_client_handle_t client, const char *u
...
         char *user_info = NULL;
         http_utils_assign_string(&user_info, url + purl.field_data[UF_USERINFO].off, purl.field_data[UF_USERINFO].len);
-        if (user_info) {
-            char *username = user_info;
-            char *password = strchr(user_info, ':');
-            if (password) {
-                *password = 0;
-                password ++;
-                http_utils_assign_string(&client->connection_info.password, password, 0);
-                HTTP_MEM_CHECK(TAG, client->connection_info.password, return ESP_ERR_NO_MEM);
```
Comment (viettrungluu) [inline components/esp_http_client/esp_http_client.c]: user_info was leaked here and below.

File: components/esp_http_client/esp_http_client.c [hunk 5]
```diff
@@ -700,37 +694,57 @@ esp_err_t esp_http_client_set_url(esp_http_client_handle_t client, const char *u
...
+            old_host = strdup(client->connection_info.host);
+            HTTP_MEM_CHECK(TAG, old_host, return ESP_ERR_NO_MEM);
+        }
+        http_utils_assign_string(&client->connection_info.host, url + purl.field_data[UF_HOST].off, purl.field_data[UF_HOST].len);
+        HTTP_MEM_CHECK(TAG, client->connection_info.host, {
+            free(old_host);
+            return ESP_ERR_NO_MEM;
+        });
+        host_was_changed = !old_host || strcasecmp(old_host, client->connection_info.host) != 0;
+        free(old_host);
```
Comment (viettrungluu) [inline components/esp_http_client/esp_http_client.c]: Note that it's OK to call free() on a null pointer.

File: components/esp_http_client/esp_http_client.c [hunk 6]
```diff
@@ -439,6 +442,21 @@ static esp_err_t esp_http_client_prepare(esp_http_client_handle_t client)
...
 }
 
+static esp_err_t set_host_header(esp_http_client_handle_t client)
+{
+    if (!client->connection_info.host)
+        return ESP_ERR_INVALID_STATE;
+    // Always append the port; we need 1 + 5 + 1 = 7 extra characters beyond the length:
+    size_t size = strlen(client->connection_info.host) + 7;
+    char *buf = (char *)malloc(size);
+    if (!buf)
```
Comment (jitin17) [inline components/esp_http_client/esp_http_client.c]: According to ESP-IDF style guide, if statement should be followed by {} braces. So, I recommend changing this to
```
    if (!buf) {
         return ESP_ERR_NO_MEM;
    }
```

File: components/esp_http_client/esp_http_client.c [hunk 7]
```diff
@@ -439,6 +442,21 @@ static esp_err_t esp_http_client_prepare(esp_http_client_handle_t client)
...
+        return ESP_ERR_NO_MEM;
+    snprintf(buf, size, "%s:%d", client->connection_info.host, client->connection_info.port);
```
Comment (jitin17) [inline components/esp_http_client/esp_http_client.c]: I think the return value of `snprintf` should be validated here.

File: components/esp_http_client/esp_http_client.c [hunk 8]
```diff
@@ -700,37 +694,57 @@ esp_err_t esp_http_client_set_url(esp_http_client_handle_t client, const char *u
...
-                *password = 0;
-                password ++;
-                http_utils_assign_string(&client->connection_info.password, password, 0);
-                HTTP_MEM_CHECK(TAG, client->connection_info.password, return ESP_ERR_NO_MEM);
-            }
-            http_utils_assign_string(&client->connection_info.username, username, 0);
-            HTTP_MEM_CHECK(TAG, client->connection_info.username, return ESP_ERR_NO_MEM);
-            free(user_info);
-        } else {
+        if (!user_info) {
```
Comment (jitin17) [inline components/esp_http_client/esp_http_client.c]: Why `user_info` has been changed to `!user_info`?

**Comment by jitin17** [comment]
@viettrungluu Left few minor comments, otherwise, this looks good to be accepted. Thanks for your efforts.

**Comment by shubhamkulkarni97** [comment]
Hi @viettrungluu, Please update this PR with suggested changes.

**Comment by mahavirj** [comment]
This was fixed in https://github.com/espressif/esp-idf/commit/b06035443fbdcd66d1db463fc8a597ca031ad0b4 (in bit different way). Thank you for contribution.

**Pull request closed without merging**